### PR TITLE
fix(atomic): use closest for internal initialization utils

### PR DESCRIPTION
**Pull request by ThibodeauJF**
https://coveord.atlassian.net/browse/KIT-2313

Some context, we used the [custom event return value](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent#return_value) as a basis for when a component has been initialized properly or not (the event’s payload is an “initialize” callback. The return value is basically:

- false if event is cancelable, and at least one of the event handlers which received event called [Event.preventDefault()](https://developer.mozilla.org/en-US/docs/Web/API/Event/preventDefault). We call prevent default from atomic-search-interface, for example/
- Otherwise true.

This is synchronous, if dealing with a simple DOM, like our basic search interfaces, this will always work. When we started to expose these custom events for our custom elements, we started to have issues with Stencil. When those events where dispatched from inside another Stencil component, they would reach the “atomic-search-interface” but with some delay, causing the dispatchEvent return value to be true. A typical race condition problem. We could argue that Stencil’s rendering inner workings is at fault, or that the method we used to make sure the component is initialized properly was not the best solution (flaky). https://github.com/coveo/ui-kit/pull/2242or we could have some sort of asynchronous polling method to check if init is called. 

TLDR: there is a false negative init error when using imbricated Stencil components.

### CODE REVIEW

File: packages/atomic/src/utils/initialization-utils.tsx [hunk 1]
```diff
@@ -144,35 +144,31 @@ export function InitializeBindings<SpecificBindings extends AnyBindings>({
...
           }
+        } catch (e) {
+          this.error = e as Error;
         }
-      );
+      });
 
-      const canceled = element.dispatchEvent(event);
-      if (canceled) {
+      if (!closest(element, initializableElements.join(', '))) {
```
Comment (ThibodeauJF) [inline packages/atomic/src/utils/initialization-utils.tsx:171]: Diff is wack, only changed this line & removed the event const.

"Closest" is used at line 44 already

**Review by louis-bompart — COMMENTED** [review]
I have a hard time understanding what we're doing here.
Could you put some flesh on the issue or PR to describe the issue and/or an example of the issue we want to solve?

Tests would be nice too :D

**Comment by btaillon** [comment]
Our events are normally supposed to bubble up through the ShadowDOM (see [`Event.composed`](https://developer.mozilla.org/en-US/docs/Web/API/Event/composed)). I'm concerned about why the event isn't getting canceled, and whether it could cause other issues.

Is this issue happening for a specific environment?

File: packages/atomic/src/utils/initialization-utils.tsx (hunk 1 above)
Comment (ThibodeauJF) [reply packages/atomic/src/utils/initialization-utils.tsx:176]: https://github.com/coveo/ui-kit/pull/2242

**Comment by ThibodeauJF** [comment]
@btaillon @louis-bompart 
Stencil screws that up, they'll intercept the event but it will bubble properly. This is why "closest" is used for the initializeBindings method already.

It's more apparent when we create some type of race condition. In the Stencil sample:

1. The results-manager component initializes as a stencil component
2. Atomic is inserted in the head and loads
3. atomic-result-list initializes and dispatches the event from inside the results-manager.
4. Stencil intercepts the event, and re-dispatches it, screwing up atomic-result-list's initialization 

If Atomic loads first, the atomic-result-list component initializes **before** results-manager is recognize as a Stencil component.

I'll add a new sanity test in the stencil sample package.

File: packages/samples/stencil/src/pages/hsp.html [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Atomic in Stencil</title>
+    <link rel="stylesheet" href="/build/atomic/themes/coveo.css" />
+    <script type="module" src="/build/app.esm.js"></script>
+
+    <script>
+      const script = document.createElement('script');
+      script.type = 'module';
+      script.src = 'https://static.cloud.coveo.com/atomic/v2/atomic.esm.js'; // fix: src="/build/atomic/atomic.esm.js"
```
Comment (ThibodeauJF) [inline packages/samples/stencil/src/pages/hsp.html:13]: switch it to see the fix in action

**Comment by btaillon** [comment]
@ThibodeauJF

> Stencil intercepts the event, and re-dispatches it, screwing up atomic-result-list's initialization

Ah, I didn't know Stencil intercepts events like this. What are the conditions for an event to get canceled? Is it canceled by the parent Stencil component when it renders custom elements it doesn't own?

> If Atomic loads first, the atomic-result-list component initializes before results-manager is recognize as a Stencil component

I'm a bit confused about how this is possible. If `results-manager` is responsible for appending `atomic-result-list` to itself (like in your PR), then I would expect that initializing Atomic first makes it see `results-manager` as being an empty element.

I get your point though.

File: packages/samples/stencil/src/pages/hsp.html [hunk 3]
```diff
@@ -0,0 +1,68 @@
+<html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Atomic in Stencil</title>
+    <link rel="stylesheet" href="/build/atomic/themes/coveo.css" />
+    <script type="module" src="/build/app.esm.js"></script>
```
Comment (ThibodeauJF) [inline packages/samples/stencil/src/pages/hsp.html:8]: Switch for `https://static.cloud.coveo.com/atomic/v2/atomic.esm.js` to see it fail

**Comment by louis-bompart** [comment]
> Intercepted was treated as cancelled
by who/what code?

**Comment by ThibodeauJF** [comment]
Deleted my older comments to avoir confusion if anyone is looking at this later one.
Added more context in the description.

**Review by btaillon — APPROVED** [review]

**Review by louis-bompart — APPROVED** [review]

File: packages/atomic/src/utils/initialization-utils.tsx [hunk 4]
```diff
@@ -171,8 +173,9 @@ export function InitializeBindings<SpecificBindings extends AnyBindings>({
         }
       );
 
-      const canceled = element.dispatchEvent(event);
-      if (canceled) {
+      element.dispatchEvent(event);
+
+      if (!closest(element, initializableElements.join(', '))) {
```
Comment (olamothe) [inline packages/atomic/src/utils/initialization-utils.tsx:178]: Do we have any kind of test for the "unhappy" path ?

ie: If I put an `atomic-facet` randomly outside of an interface, it should throw an error.

**Review by olamothe — APPROVED** [review]

File: packages/atomic/src/utils/initialization-utils.tsx (hunk 4 above)
Comment (ThibodeauJF) [reply packages/atomic/src/utils/initialization-utils.tsx:178]: Yes it's covered
